presto-native-execution/presto_cpp/main/expression/RowExpressionOptimizer.cpp

```diff
@@ -0,0 +1,709 @@
+  res["arguments"] = json::array();
+
+  // Arguments for switch expression include special form expression 'when'
+  // so it is constructed separately. If the switch expression evaluation found
+  // a case that always evaluates to true, the second value in pair switchResult
+  // will be false and the first value in pair will contain the value
+  // corresponding to the simplified case.
+  if (form == "SWITCH") {
+    auto switchResult = getSwitchSpecialFormArgs(expr, input);
+    if (switchResult.second) {
+      res["arguments"] = switchResult.first;
+    } else {
+      return switchResult.first.front();
+    }
+  } else {
+    json::array_t inputArguments = input["arguments"];
+    const auto numInputs = exprInputs.size();
+    VELOX_USER_CHECK_LE(numInputs, inputArguments.size());
+    for (auto i = 0; i < numInputs; i++) {
+      res["arguments"].push_back(
+          veloxExprToRowExpression(exprInputs[i], inputArguments[i]));
+    }
+  }
+  res["returnType"] = getTypeSignature(expr->type());
+
+  return res;
+}
+
+json RowExpressionConverter::toConstantRowExpression(
+    const exec::ExprPtr& expr) {
+  json res;
+  auto constantExpr = std::dynamic_pointer_cast<const exec::ConstantExpr>(expr);
+  VELOX_USER_CHECK_NOT_NULL(constantExpr);
+  auto constantRowExpr = getConstantRowExpression(constantExpr);
+  protocol::to_json(res, constantRowExpr);
+  return res;
+}
+
+json RowExpressionConverter::toCallRowExpression(
+    const exec::ExprPtr& expr,
+    const json& input) {
+  json res;
+  res["@type"] = "call";
+  protocol::Signature signature;
+  std::string exprName = expr->name();
+  if (veloxToPrestoOperatorMap_.find(expr->name()) !=
+      veloxToPrestoOperatorMap_.end()) {
+    exprName = veloxToPrestoOperatorMap_.at(expr->name());
+  }
+  signature.name = exprName;
+  res["displayName"] = exprName;
+  signature.kind = protocol::FunctionKind::SCALAR;
+  signature.typeVariableConstraints = {};
+  signature.longVariableConstraints = {};
+  signature.returnType = getTypeSignature(expr->type());
+
+  std::vector<protocol::TypeSignature> argumentTypes;
+  auto exprInputs = expr->inputs();
+  auto numArgs = exprInputs.size();
+  argumentTypes.reserve(numArgs);
+  for (auto i = 0; i < numArgs; i++) {
+    argumentTypes.emplace_back(getTypeSignature(exprInputs[i]->type()));
+  }
+  signature.argumentTypes = argumentTypes;
+  signature.variableArity = false;
+
+  protocol::BuiltInFunctionHandle builtInFunctionHandle;
+  builtInFunctionHandle._type = "$static";
+  builtInFunctionHandle.signature = signature;
+  res["functionHandle"] = builtInFunctionHandle;
+  res["returnType"] = getTypeSignature(expr->type());
+  res["arguments"] = json::array();
+  for (const auto& exprInput : exprInputs) {
+    res["arguments"].push_back(veloxExprToRowExpression(exprInput, input));
+  }
+
+  return res;
+}
+
+json RowExpressionConverter::veloxExprToRowExpression(
+    const exec::ExprPtr& expr,
+    const json& input) {
+  if (expr->type()->isRow()) {
```

Comment:
Is there some reason for this to be handled right in the beginning of this function ? Handling constant, field reference, then Call expression seems a more intuitive order.